language-plutus-core/src/Language/PlutusCore/TypeRenamer.hs

```diff
@@ -1,120 +1,182 @@
+    where st' = IM.fromList (zip keys keys)
+          keys = IM.keys st
+          m = fst (IM.findMax st)
+
+newtype Identifiers = Identifiers { _identifiers :: IM.IntMap Int }
+    deriving Show
+
+type MaxM = State Int
+
+identifiers :: Lens' Identifiers (IM.IntMap Int)
+identifiers f s = fmap (\x -> s { _identifiers = x }) (f (_identifiers s))
+
+modifyIdentifiers :: Int -> Int -> Identifiers -> Identifiers
+modifyIdentifiers u m = over identifiers (IM.insert u (m+1) . IM.insert (m+1) (m+1))
+
+lookupId :: Int -> Identifiers -> Maybe Int
+lookupId u st = IM.lookup u (_identifiers st)
+
+-- this convoluted affair lets us track the maximum in a global state monad,
+-- while keeping the table for renaming local (so that we don't rename things in
+-- function applications)
+renameTerm :: Identifiers -> Term TyName Name a -> MaxM (Term TyName Name a)
+renameTerm st t@(LamAbs x (Name x' s (Unique u)) ty t') = do
+    m <- get
+    let st' = modifyIdentifiers u m st
+        pastDef = lookupId u st
     case pastDef of
-        Just _ -> LamAbs x (Name x' s (Unique $ m+1)) <$> rewriteWith (Unique u) (Unique $ m+1) t'
+        Just _ ->
+            modify (+1) >>
+            LamAbs x (Name x' s (Unique (m+1))) <$> renameType st' ty <*> renameTerm st' t'
         _      -> pure t
-renameTerm (TyInst x t tys) = TyInst x <$> renameTerm t <*> traverse renameType tys
-renameTerm (Apply x t ts)   = Apply x <$> renameTerm t <*> traverse renameTerm ts
-renameTerm (Unwrap x t)     = Unwrap x <$> renameTerm t
-renameTerm x                = pure x
-
-cataM :: (Recursive t, Traversable (Base t), Monad m) => (Base t a -> m a) -> (t -> m a)
-cataM phi = c where c = phi <=< (traverse c . project)
-
--- rename a particular unique in a subterm
-rewriteWith :: Unique -> Unique -> Term a -> IdentifierM (Term a)
-rewriteWith i j@(Unique u) = cataM aM where
-    aM (VarF x (Name x' s i')) | i == i' = do
-        modify (first (IM.insert u s))
-        pure $ Var x (Name x' s j)
-    aM (LamAbsF x (Name x' s i') t) | i == i' = do
-        modify (first (IM.insert u s))
-        pure $ LamAbs x (Name x' s j) t
-    aM x = pure (embed x)
-
--- TODO do the same thing here
-renameType :: Type a -> IdentifierM (Type a)
-renameType = pure
+renameTerm st t@(Fix x (Name x' s (Unique u)) ty t') = do
+    m <- get
+    let st' = modifyIdentifiers u m st
+        pastDef = lookupId u st
+    case pastDef of
+        Just _ ->
+            modify (+1) >>
+            Fix x (Name x' s (Unique (m+1))) <$> renameType st' ty <*> renameTerm st' t'
+        _ -> pure t
+renameTerm st t@(Var x (Name x' s (Unique u))) =
+    case pastDef of
+        Just j -> pure $ Var x (Name x' s (Unique j))
+        _      -> pure t
+    where pastDef = lookupId u st
+renameTerm st (Apply x t ts) = Apply x <$> renameTerm st t <*> traverse (renameTerm st) ts
+renameTerm st (Unwrap x t) = Unwrap x <$> renameTerm st t
+renameTerm _ x@Constant{} = pure x
+renameTerm _ x = pure x
```

Comment:
Please don't use catch all cases. Instead, let's enumerate all the variants. Then, when we change the datatype definition, we'll get a warning if we don't extend all functions operating on the datatype as well.

Reply:
Also, don't we need to cover `TyAbs` and `Wrap`? (Or maybe you just wanted my feedback on the code so far. That's fine, of course, but maybe good to add a comment, just so that I know.)